File: parser/datacap-parser-trino/src/main/java/io/edurt/datacap/parser/node/Node.java
package io.edurt.datacap.parser.node;

import io.edurt.datacap.parser.ast.AstVisitor;

import java.util.List;
import java.util.Optional;

import static java.util.Objects.requireNonNull;

public abstract class Node
{
    private final Optional<NodeLocation> location;

    protected Node(Optional<NodeLocation> location)
    {
        this.location = requireNonNull(location, "location is null");
    }

    protected static boolean sameClass(Node left, Node right)
    {
        if (left == right) {
            return true;
        }
        return left.getClass() == right.getClass();
    }

    /**
     * Accessible for {@link AstVisitor}, use {@link AstVisitor#process(Node, Object)} instead.
     */
    public <R, C> R accept(AstVisitor<R, C> visitor, C context)
    {
        return visitor.visitNode(this, context);
    }

    public Optional<NodeLocation> getLocation()
    {
        return location;
    }

    public abstract List<? extends Node> getChildren();

    // Force subclasses to have a proper equals and hashcode implementation
    @Override
    public abstract int hashCode();

    @Override
    public abstract boolean equals(Object obj);

    @Override
    public abstract String toString();

    /**
     * Compare with another node by considering internal state excluding any Node returned by getChildren()
     */
    public boolean shallowEquals(Node other)
    {
        throw new UnsupportedOperationException("not yet implemented: " + getClass().getName());
    }
}

File: parser/datacap-parser-mysql/src/main/kotlin/io/edurt/datacap/parser/mysql/companion/ast/Node.kt
package io.edurt.datacap.parser.mysql.companion.ast

import io.edurt.datacap.parser.mysql.companion.AstVisitor
import java.util.*

abstract class Node(location: Optional<NodeLocation>) {
    open val location: Optional<NodeLocation> = Objects.requireNonNull(location, "location is null")

    protected open fun sameClass(left: Node, right: Node): Boolean {
        if (left == right) {
            return true
        }
        return left.javaClass == right.javaClass
    }

    open fun <R, C> accept(visitor: AstVisitor<R, C>, context: C): R? {
        return visitor.visitNode(this, context)
    }

    protected open fun shallowEquals(other: Node): Boolean {
        throw UnsupportedOperationException("not yet implemented: " + javaClass.name)
    }
}
